src/utils/rlp.cairo

```diff
@@ -1,69 +1,133 @@
+        return (1, 1 + len_bytes_count, list_len);
+    }
+
+    // @notice Decodes a Recursive Length Prefix (RLP) encoded data.
+    // @notice This function decodes the RLP encoded data into a list of items.
+    // Each item is a struct containing the length of the data, the data itself, and a flag indicating whether the data is a list.
+    // The function first determines the type of the RLP data (string or list) and then processes it accordingly.
+    // If the data is a string, it is simply added to the items.
+    // If the data is a list, it is recursively decoded.
+    // After processing the first item, the function checks if there is more data to decode and if so,
+    // it recursively decodes the remaining data and adds the decoded items to the list of items.
+    // @param data_len The length of the data to decode.
+    // @param data The RLP encoded data.
+    // @return items_len The number of items decoded.
+    // @return items The decoded RLP items.
+    func decode{range_check_ptr, items_len: felt, items: Item*}(data_len: felt, data: felt*) {
+        alloc_locals;
+
+        if (data_len == 0) {
+            return ();
+        }
+
+        let (rlp_type, offset, len) = decode_type(data_len=data_len, data=data);
+
+        if (rlp_type == 0) {
+            // Case string
+            if (len == 0) {
+                let (empty_array: felt*) = alloc();
+                tempvar rlp_string = Item(data_len=0, data=empty_array, is_list=0);
+                assert [items + items_len] = rlp_string;
+                tempvar items_len = items_len + 1 * Item.SIZE;
+                tempvar range_check_ptr = range_check_ptr;
+            } else {
+                let rlp_string = Item(data_len=len, data=data + offset, is_list=0);
+                assert [items + items_len] = rlp_string;
+                tempvar items_len = items_len + 1 * Item.SIZE;
+                tempvar range_check_ptr = range_check_ptr;
+            }
+        } else {
+            // Case List
+            if (len == 0) {
+                let (empty_list: Item*) = alloc();
+                tempvar rlp_list = Item(data_len=0, data=cast(empty_list, felt*), is_list=1);
+                assert [items + items_len] = rlp_list;
+                tempvar items_len = items_len + 1 * Item.SIZE;
+                tempvar range_check_ptr = range_check_ptr;
+            } else {
+                let sub_items_len = 0;
+                let (sub_items: Item*) = alloc();
+                decode{items_len=sub_items_len, items=sub_items}(data_len=len, data=data + offset);
+                tempvar rlp_list = Item(
+                    data_len=sub_items_len, data=cast(sub_items, felt*), is_list=1
+                );
+                assert [items + items_len] = rlp_list;
+                tempvar items_len = items_len + 1 * Item.SIZE;
+                tempvar range_check_ptr = range_check_ptr;
+            }
+        }
```

Comment:
it looks like if you invert the conditions order, you can save because both cases with `len=0` would be 

```
let (empty: felt*) = alloc();
let item = Item(data_len=0, data=empty, is_list=rlp_type);
assert [items + items_len] = rlp_list;
tempvar items_len = items_len + 1 * Item.SIZE;
tempvar range_check_ptr = range_check_ptr;
```